### DISCUSSION TITLE
Hosting in a sub-directory?

### DISCUSSION THREAD

**Discussion by murty2** [DustinBrett/daedalOS#449] [Q&A] [answered]
The final output in out/ directory assumes that I host files in the root of my website under /. It contains many absolute URL links starting with /
But I want to place the output files in /sub not / of my website. How can I configure such that relative URLs are generated everywhere (preferred), and if that's not possible, URLs that start with /sub/_next/...  etc?

**Comment by DustinBrett** [comment]
I don't think this would be easy to do because lots of the logic is based around starting at the root path. Perhaps if you can, put daedalOS on a subdomain of your site, such as, `os.yourdomain.com` so it's still at the root of it's path. If you wanted to have it work in a sub path I am not even sure where you'd start. Maybe just move it there and then whack-a-mole the errors.

**Reply by murty2** [reply]
I wanted to serve via sub directory so I don't have to worry about getting another DNS entry. I also thought about ProxyPass and ProxyPassReverse connections to backend yarn serve, but Its an overkill. I did manage to make it work with the following changes. This I'm sure sure isn't sufficient if I leave all the features enabled
```
# WARNING: Not showing icons for Short Cuts under: My PC -> Users -> Public -> Start Menu

cd /var/www/html/webtop || exit

for i in *
do
  find . -type f \( -iname '*.js' -o -iname '*.json' -o -iname '*.css' -o -iname '*.html' \) | while read FILE
  do
    [[ "${FILE,,}" =~ vanta || "$FILE" == */slideshow\.json ]] && continue
    grep -q /webtop/ "$FILE" || sed -zi "s/\/\($i\)/\1/g" "$FILE"
  done
done

grep -q /webtop/ Users/Public/Pictures/slideshow.json ||
sed -i 's|"/Users|"/webtop/Users|g' Users/Public/Pictures/slideshow.json

find . -type f -name '*.url' | while read FILE
do
  grep -q /webtop/ "$FILE" || sed -i 's|=/|=/webtop/|g' "$FILE"
done

grep -rli System/Vanta | while read FILE
do
  grep -q /webtop/ "$FILE" || sed -zi 's|System/Vanta|/webtop/System/Vanta|gI' "$FILE"
done

find . -type f -exec chmod 644 {} \;
find . -type d -exec chmod 755 {} \;
```

**Reply by DustinBrett** [reply]
Ya I am not sure about doing things this way because a lot of parts of the code make assumptions based on being at root, it's used inherently in the code to make it feel more like a file system where `/` is called via HTTP.

**Reply by murty2** [reply]
I think webpack also needs some configuration to set baseurl; I dont know. This is what I did as shown below. This works but everything stops working if I access /sub/?app=Browser

1. Added basePath like you did. I did not add assetPrefix. But I also added distDir: "out/sub" above compiler:

2. Added "/sub" to SYSTEM_PATHS in utils/constant.js

3. Run: yarn build

4. Run the following fix
```
SUBDIR=sub

for X in .index System 'Program Files'
do
  grep -rl "[^a-zA-Z0-9]/$X\b" out |
  while read FILE
  do
    sed -zi "s|\([^a-zA-Z0-9]\)\(/$X\b\)|\1/$SUBDIR\2|g" "$FILE"
  done
done

sed -i "s|\([^a-zA-Z0-9]\)/Users|\1/$SUBDIR/Users|g" out/$SUBDIR/Users/Public/Pictures/slideshow.json
```

**Comment by murty2** [comment]
While investigating a better and more robust solution, I discovered that many Ajax requests are made without HTTP-Referer header :(

**Please see if you can fix this, such that all http and https requests, no matter how they are made, are made with HTTP-Referer header.**

I place my final out/* contents in my DocRoot/webtop. The following Nginx configuration will rewrite requests made to / to /webtop
```
    location / {
      if ($http_referer ~ /webtop(?:/|\?|$)) {
        return 301 $scheme://$host/webtop$request_uri;
      }
      location ~ ^/(?:System|Users|Program\sFiles|Program%20Files)(?:/|\?|$) {
        return 301 $scheme://$host/webtop$request_uri;
      }
      location = /screenshot.png {
        return 301 $scheme://$host/webtop$request_uri;
      }
      location /webtop {
        # If you don't have any directives for /webtop/... URLs, comment the following. But do not delete this location block
        include /etc/nginx/conf.d/Files;
      }
    }
```

The drawback of this method is that when some of the Ajax requests are made **without** HTTP-Referer header for /System/... or /Users/... we redirect the browser to /webtop. But then I cannot have two installations of daedalOS in two different sub-directories: /webtop and /webtopnew

This is why it is important to modify the code such that ALL requests are made with HTTP-Referer header. I realize that HTTP-Referer can be spoofed and not reliable... but right now, I don't have any better solution

**Comment by murty2** [comment, marked as answer]
Here's the full working solution. It's server-side but it works. Suppose you want to serve daedalOS as example.com/webtop 

After you run "yarn build" it will generate "out" directory.

- Copy out directory: cp -a out /var/www/html/webtop
- Run the following
```
find /var/www/html/webtop -type f -exec chmod 644 {} \;
find /var/www/html/webtop -type d -exec chmod 755 {} \;
```
- Replace no-referrer by same-origin to make sure that XMLHttpRequests always send HTTP-Referer header
```
grep -rlw no-referrer /var/www/html/webtop |
while read FILE
do
  sed -i 's/\bno-referrer\b/same-origin/g' "$FILE"
done
```
- Configure Nginx web server for domain.com. Put this in /etc/nginx/conf.d/domain.conf
```
    location / {
      if ($http_referer ~ /webtop(?:/|\?|$)) {
        return 301 $scheme://$host/webtop$request_uri;
      }
      location ~ ^/webtop(?:/|$) {
       # Usual configuration directives. If none, just leave this location block empty
      }
    }
```
- Configuration for Apache and Litespeed. Put this in {DocRoot}/.htaccess
```
<IfModule mod_rewrite.c>
  RewriteCond %{REQUEST_URI} !^/webtop(?:/|$)
  RewriteCond %{HTTP_REFERER} /webtop(?:/|\?|$)
  RewriteRule ^ /webtop%{REQUEST_URI} [END,R=301]
</IfModule>
```

The above Nginx configuration says: If the request URI starts with /webtop do usual stuff or nothing special, to send response. If not, check its Referer. If the Referer starts with /webtop that means the request is for daedalOS and therefore redirect it to /webtop sub-directory. We only check the Referer if the URI does not start with /webtop, because those requests can be fulfilled normally without any rewriting or redirecting.
